packages/edit-post/src/components/header/index.js

```diff
@@ -29,10 +35,22 @@ function Header( {
 	isPublishSidebarEnabled,
 	togglePublishSidebar,
 	hasActiveMetaboxes,
+	hasPublishAction,
 	isSaving,
+	isPublished,
+	isScheduled,
+	isBeingScheduled,
+	isPending,
+	isSmallViewport,
 } ) {
 	const toggleGeneralSidebar = isEditorSidebarOpened ? closeGeneralSidebar : openGeneralSidebar;
 
+	const shouldShowButton = (
+		( ! isPublishSidebarEnabled && ! isSmallViewport ) ||
+		isPublished ||
+		( isScheduled && isBeingScheduled ) ||
+		( isPending && ! hasPublishAction && ! isSmallViewport )
+	);
```

Comment:
I'm having a hard time understanding this expression. Could we refactor it as a function which uses early `return`s? Exporting said function would then let us write unit tests to test the logic.